articles/content_filtering.md

```diff
@@ -0,0 +1,222 @@
+
+<img src="../img/content_filter/action_sequence_with_filter.png">
+
+#### Parameter Events
+
+There are two types of parameter events subscription, one is for system(TimeSource) which is controlled and taken care by rclcpp and rclpy internally. And the other is user application to register user callback to be fired when parameter events come in.
+
+It would be good to create ContentFilteredTopic for each of them. This can make responsibility clear and simple, can avoid the complication about filtering expression merge between user filtering and system filtering.
+
+|  ContentFilterTopic  |  Description  |
+| :--- | :--- |
+|  /parameter_events_system_Node_ID  |  parameter events ContentFilteredTopic for system internal usage, Node ID would be uuid to identify this NodeBase. (same with user uuid)  |
+|  /parameter_events_user_Node_ID  |  parameter events ContentFilteredTopic for user callback  usage, Node ID would be uuid to identify this NodeBase. (same with system uuid)  |
+
+(*) These ContentFilteredTopic should not show up to user. (e.g ros2 topic list)
+
+- TimeSource
+  **use_sim_time** event on self-node_base is subscribed internally to check if **use_sim_time** parameter is changed or not. so internally this **use_sim_time** AND self node name always must be in filter_expression and expression_parameters via ContentFilteredTopic if it is supported. see [here](https://github.com/ros2/rclcpp/blob/99286978f92c30fe171313bf0785d6b6272c3257/rclcpp/src/rclcpp/time_source.cpp#L123-L125).
+  **use_sim_time** parameter event must be guaranteed by system to keep that in the filter and parameter expression when using ContentFilteredTopic.
+
+- User frontend (rclcpp/rclpy)
+  User API will be added to manage filtering configuration, so that user application can set its own filter_expression and expression_parameters. Also using AsyncParametersClient::on_parameter_event, user can take care of the parameter event with user callback only for filtered parameter events.
+  **[T.B.D] Filtering should be compatible for user even if rmw_implementation does not support ContentFilteredTopic, this can be done with using ParameterEventsFilter. But the problem here is ParameterEventsFilter only filters parameter name and event type such as NEW, DELETED, CHANGED. there is a huge gap of the flexibility that dds filtering expression and parameter expression possesses. This is what we need to discuss.**
+
+<img src="../img/content_filter/parameter_event_sequence_with_filter.png">
+
+### Interfaces
+
+#### rcl
+
+- add API `is_cft_supported` to allow frontend(rclcpp/rclpy) can know CFT is supported or not.
+  e.g) `rcl_subscription_is_cft_supported` to access `is_cft_supported`.
+- add `rcl_create_cft_subscription` to call `rmw_create_cft_subscription`.
+- add `rcl_cft_set/get_expression_parameters` to call `rmw_cft_set/get_expression_parameters`.

```

Comment:
Having a `rcl_is_cft_supported` function sounds like a good idea, but I don't get why we need all the others.

A cft expression in the subscription options sounds like a better idea than a special function to create a cft_subscription.
The `cft` expression can be a custom object, and we can add a proper API to configure it.

Reply:
user application is likely to change filtering expression and parameter expressions at runtime, so i think that we need to provide methods to do that dynamically.

Reply:
> user application is likely to change filtering expression and parameter expressions at runtime, so i think that we need to provide methods to do that dynamically.

Do you mean changing the expression in an already created subscription?
Or dynamically creating an expression?

In the first case, having a special object for the `cft` subscription makes sense.
In the second, only having a object for the expression makes sense.

Reply:
I am not sure if i follow you precisely,

> Do you mean changing the expression in an already created subscription?

I say yes, this is needed especially for action topics. filtering expression and expression parameters for goal id need to be adjusted at runtime.

Reply:
Oh yeah, we need to be able to modify the filtering expression of an already created cft subscription (or at least, the expression parameters).